CASSANDRA-19815: Decouple Cassandra types from Spark types so Cassandra types can be u…

jberragan opened the pull request:
…sed independently from Spark

jberragan committed:
Fix code style and jdk8 compilation

yifan-c commented on cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/data/CqlField.java:
```diff
@@ -35,19 +35,16 @@
 {
     private static final long serialVersionUID = 42L;
 
+    private static final Comparator<String> STRING_COMPARATOR = String::compareTo;
```
Only `STRING_COMPARATOR` is `private`.

yifan-c commented on cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/data/CqlField.java:
```diff
@@ -201,6 +184,11 @@ public interface CqlCollection extends CqlType
 
         CqlField.CqlType type(int position);
+
+        default boolean isComplex()
```
add javadoc

yifan-c commented on cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/data/CqlField.java:
```diff
@@ -222,12 +210,17 @@ public interface CqlTuple extends CqlCollection
         ByteBuffer serializeTuple(Object[] values);
 
-        Object[] deserializeTuple(ByteBuffer buffer, boolean isFrozen);
+        Object[] deserializeTuple(TypeConverter converter, ByteBuffer buffer, boolean isFrozen);
```
Please add javadoc for the API methods touched in this patch.

yifan-c commented on cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/data/CqlField.java:
```diff
@@ -330,14 +329,39 @@ public CqlType type()
+        return type().deserializeToType(converter, buffer, isFrozen);
+    }
+
+    public Object deserializeToJava(ByteBuffer buffer)
```
nit: `deserializeToJavaType`

yifan-c commented on cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/data/TypeConverter.java:
```diff
@@ -0,0 +1,40 @@
+
+import org.jetbrains.annotations.NotNull;
+
+public interface TypeConverter
```
how about `SparkTypeConverter`?

jberragan replied:
`TypeConverter` is designed to be generic and de-coupled from Spark. `SparkSqlTypeConverter` extends `TypeConverter`.

yifan-c commented on cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/data/TypeConverter.java:
```diff
@@ -0,0 +1,40 @@
+public interface TypeConverter
+{
+    /**
+     * Converts Cassandra native value to desired equivalent.
```
not sure what does it mean with "Cassandra native value". Is the `Object value` indeed java type value?

yifan-c commented on cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/data/CqlField.java:
```diff
@@ -330,14 +329,39 @@ public CqlType type()
+     * @param isFrozen  true if the Cassandra type is frozen
+     * @return deserialized object converted to custom type.
+     */
+    public Object deserializeToType(TypeConverter converter, ByteBuffer buffer, boolean isFrozen)
```
nit: `deserializeToSparkType` to better distinguish?

jberragan replied:
`TypeConverter` is designed to be generic and not tied to Spark, so it could be some other type.

yifan-c commented on cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/data/CqlField.java:
```diff
@@ -473,6 +462,11 @@ public void write(Kryo kryo, Output output, CqlField field)
+        return notImplemented(type.toString());
+    }
+
+    public static UnsupportedOperationException notImplemented(String type)
+    {
+        return new UnsupportedOperationException(type + " type not implemented");
     }
```
The exception message is very much about type only. Can you make it a `private` method? To avoid being used in the other context.

jberragan replied:
It is used in `NotImplementedTraits` and some tests which is why it is `public`, I can re-work if it's a blocker.

yifan-c replied:
Not a blocker. Just want to avoid any confusion. As long as they are used in the context of "type" only, it is good.

yifan-c commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/sparksql/SparkCellIterator.java:
```diff
@@ -325,7 +329,10 @@ private void maybeRebuildClusteringKeys(@NotNull ByteBuffer columnNameBuf)
             Object newObj = deserialize(field, ByteBufferUtils.extractComponent(columnNameBuf, index++));
             Object oldObj = values[field.position()];
-            if (newRow || oldObj == null || newObj == null || !field.equals(newObj, oldObj))
+            // Historically, we compare equality of clustering keys using the Spark types
+            // to determine if we have moved to a new 'row'. We could also compare using the Cassandra types
+            // or the raw ByteBuffers before converting to Spark types.
+            if (newRow || oldObj == null || newObj == null || !sparkSqlTypeConverter.toSparkType(field.type()).equals(newObj, oldObj))
```
Question: is the comment a todo (for moving to compare using Cassandra types/raw bytes)?

jberragan replied:
Not necessarily, comparing with Spark types has worked fine for now. Comparing the raw bytes could be a bit faster. It is just an assumption that was intrinsic but not fully called out.

yifan-c reviewed (commented):
submit my review so far. I will continue the review in the other day.

yifan-c commented:
This flaky test `org.apache.cassandra.spark.EndToEndTests#testLargeBlobExclude` might be related. https://app.circleci.com/pipelines/github/yifan-c/cassandra-analytics/92/workflows/733d92a6-613e-4a00-ac3e-406111219d8a/jobs/853/tests

jberragan committed:
Review comments

jberragan committed:
Fix tuple tests

jberragan committed:
Fix DataTypeSerializationTests.testTuples

jberragan commented:
Green CI: https://app.circleci.com/pipelines/github/jberragan/cassandra-analytics/64/workflows/3eebf8bb-d4f2-43dc-ae0b-325b4772543e

jberragan committed:
Add CHANGES.txt entry

jberragan committed:
Rename method
Add back removed code

jberragan committed:
Minor comment and code clean up

jberragan committed:
Decouple Cassandra types from Spark types so Cassandra types can be used independently from Spark

jberragan committed:
Fix checkstyle

jberragan committed:
Update comment

jberragan commented:
Latest CircleCI: https://app.circleci.com/pipelines/github/jberragan/cassandra-analytics/71/workflows/9235a1e4-b44f-4c91-8691-ca92f70a9122

jberragan commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/sparksql/SparkCellIterator.java:
```diff
@@ -325,7 +329,10 @@ private void maybeRebuildClusteringKeys(@NotNull ByteBuffer columnNameBuf)
+            // Historically, we compare equality of clustering keys using the Spark types
+            // to determine if we have moved to a new 'row'. We could also compare using the Cassandra types
+            // or the raw ByteBuffers before converting to Spark types  - this mightb be slightly more performant.
+            if (newRow || oldObj == null || newObj == null || !sparkSqlTypeConverter.toSparkType(field.type()).equals(newObj, oldObj))
```
TODO: convert CqlType to SparkType once in the constructor and re-use.

jberragan committed:
Move CassandraSchema and SchemaBuilder into cassandra-four-zero-types to be accessible without Spark dependency

jberragan committed:
Initialize SparkTypes in SparkCellIterator once to avoid converting in the hot path

jberragan commented:
Latest CircleCI: https://app.circleci.com/pipelines/github/jberragan/cassandra-analytics/73/workflows/3f5e406a-98cd-44e0-9b2a-7001975742e6

yifan-c commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/BinaryTraits.java:
```diff
@@ -0,0 +1,71 @@
+import org.apache.spark.sql.types.DataTypes;
+import org.jetbrains.annotations.NotNull;
+
+public interface BinaryTraits extends SparkType
```
I was confused by the name. At the first sight, I thought it was a `scala` source code, since `trait` is the scala's interface. 
Can you please not use the word `trait` for java interface? This patch add many traits.

jberragan replied:
I'll go ahead and remove it but I think trait or mixin is a common term not necessarily tied to any particular language

yifan-c replied:
Thank you

yifan-c commented on cassandra-analytics-core/src/test/java/org/apache/cassandra/spark/data/CqlFieldComparators.java:
```diff
@@ -38,52 +39,52 @@
 public class CqlFieldComparators extends VersionRunner
 {
 
```
Test class not named with "Test"

yifan-c commented on cassandra-analytics-core/src/test/java/org/apache/cassandra/spark/reader/DataTypeSerializationTests.java:
```diff
@@ -152,151 +153,153 @@ public void testUUID()
+                                                                                                    .serialize(BigDecimal.valueOf(Integer.MAX_VALUE))));
+            assertEquals(Decimal.apply(Integer.MIN_VALUE), bridge.decimal().deserializeToType(bridge.typeConverter(),
+                                                                                              bridge.decimal()
+                                                                                                    .serialize(BigDecimal.valueOf(Integer.MIN_VALUE))));
```
not easy to read. Can you extract a helper method?  For example, 

```
    Object toDecimal(CassandraBridge bridge, long value)
    {
        return bridge.decimal().deserializeToType(bridge.typeConverter(),
                                                  bridge.decimal().serialize(BigDecimal.valueOf(value)));
    }
```


If it looks good, can you update the other places with the same pattern?

jberragan replied:
I made the update here: https://github.com/apache/cassandra-analytics/pull/71/commits/b47f6fe3d4d04a1b77a5a6f2635c9d599a70c488

yifan-c commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/SparkSqlTypeConverter.java:
```diff
@@ -0,0 +1,73 @@
+    {
+        return toSparkType(cqlType).toTestRowType(value);
+    }
+}
```
javadoc for all methods and the interface.

yifan-c commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/Ascii.java:
```diff
@@ -0,0 +1,30 @@
+
+public class Ascii implements StringTraits
+{
+    public static final org.apache.cassandra.spark.data.converter.types.Ascii INSTANCE = new org.apache.cassandra.spark.data.converter.types.Ascii();
```
The class qualifier is redundant. 

```suggestion
    public static final Ascii INSTANCE = new Ascii();
```

yifan-c commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/BigInt.java:
```diff
@@ -0,0 +1,30 @@
+
+public class BigInt implements LongTraits
+{
+    public static final org.apache.cassandra.spark.data.converter.types.BigInt INSTANCE = new org.apache.cassandra.spark.data.converter.types.BigInt();
```
redundant class qualifier

yifan-c commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/Blob.java:
```diff
@@ -0,0 +1,30 @@
+
+public class Blob implements BinaryTraits
+{
+    public static final org.apache.cassandra.spark.data.converter.types.Blob INSTANCE = new org.apache.cassandra.spark.data.converter.types.Blob();
```
and here, and many other classes in the package

yifan-c commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/Boolean.java:
```diff
@@ -0,0 +1,60 @@
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+
+public class Boolean implements SparkType
```
I would avoid naming it as `Boolean`. At the callsites, it only causes confusion with `java.lang.Boolean`

yifan-c commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/Double.java:
```diff
@@ -0,0 +1,61 @@
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+
+public class Double implements SparkType
```
Please rename the class.

yifan-c commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/Float.java:
```diff
@@ -0,0 +1,61 @@
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+
+public class Float implements SparkType
```
Rename this class to avoid conflict with `java.lang.Float`

yifan-c commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/Duration.java:
```diff
@@ -0,0 +1,30 @@
+
+package org.apache.cassandra.spark.data.converter.types;
+
+public class Duration implements NotImplementedTraits
```
It potentially causes confusion with `java.time.Duration`, but I am not strong on renaming it.

yifan-c commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/complex/List.java:
```diff
@@ -0,0 +1,45 @@
+import org.apache.cassandra.spark.data.CqlField;
+import org.apache.cassandra.spark.data.converter.SparkSqlTypeConverter;
+
+public class List implements ListTrait
```
Not a fan of the name. To easy to be confused with `java.util.List`. Same for "Map" and "Set"

yifan-c reviewed (changes requested):
Submit my review so far. Mostly on code style / naming

jberragan committed:
Rename Spark types
Rename CqlFieldComparatorTests

jberragan committed:
Merge branch 'trunk' into CASSANDRA-19815

jberragan committed:
Add java docs for all SparkSqlTypeConverter and SparkType methods

jberragan committed:
Clean up and simplify DataTypeSerializationTests

yifan-c commented on cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/data/TypeConverter.java:
```diff
@@ -0,0 +1,42 @@
+
+public interface TypeConverter
+{
+    default Object convert(CqlField.CqlType cqlType, @NotNull Object value)
+    {
+        return convert(cqlType, value, false);
+    }
```
This method is only used for test. Can we remove it?

jberragan replied:
ok

yifan-c commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/complex/SparkTuple.java:
```diff
@@ -0,0 +1,137 @@
+        }
+        else
+        {
+            Object[] array = (Object[]) value;
```
nit: How about checking the value type is indeed `Object[]`? And throw IllegalArgumentException, instead of cast error. Maybe there could be some util method to cast types with error message that provides better context on error.

yifan-c commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/complex/ListFeatures.java:
```diff
@@ -0,0 +1,102 @@
+import org.jetbrains.annotations.NotNull;
+import scala.collection.mutable.WrappedArray;
+
+interface ListFeatures extends CollectionFeatures
```
why interface? Would abstract class be a better fit?

jberragan replied:
I can probably move these into `SparkList` as only `SparkList` implements (and `SparkSet` extends `SparkList`)

yifan-c commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/complex/ListFeatures.java:
```diff
@@ -0,0 +1,102 @@
+    @Override
+    default Object toTestRowType(Object value)
+    {
+        return Stream.of((Object[]) ((WrappedArray<Object>) value).array())
```
Can you add the explain on why the redundant cast is required? Since `array()` returns `Object[]` already.

jberragan replied:
`array()` returns `Object`

yifan-c replied:
Interesting.. I can see that `def array: Array[T]` returns arrays..

yifan-c commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/complex/SparkUdt.java:
```diff
@@ -0,0 +1,172 @@
+        }
+        else
+        {
+            return udtToSparkSqlType((java.util.Map<String, Object>) value);
```
nit: add a comment to explain where the `Map<String, Object>` is coming from. (org.apache.cassandra.spark.data.complex.CqlUdt#deserializeUdt)

I do not think the full qualified class name is required for map.

yifan-c reviewed (commented):
Sorry that the review has taken a long time. I think those are my final comments.

jberragan committed:
More review comments:
- remove unnecessary convert method
- remove unnecessary ListFeatures interface
- add instanceof checks before casting
- more comments

jberragan committed:
Merge branch 'trunk' into CASSANDRA-19815

frankgh commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/sparksql/SparkCellIterator.java:
```diff
@@ -325,7 +337,10 @@ private void maybeRebuildClusteringKeys(@NotNull ByteBuffer columnNameBuf)
-            if (newRow || oldObj == null || newObj == null || !field.equals(newObj, oldObj))
+            // Historically, we compare equality of clustering keys using the Spark types
+            // to determine if we have moved to a new 'row'. We could also compare using the Cassandra types
+            // or the raw ByteBuffers before converting to Spark types  - this mightb be slightly more performant.
```
```suggestion
            // or the raw ByteBuffers before converting to Spark types  - this might be slightly more performant.
```

frankgh commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/SparkDate.java:
```diff
@@ -0,0 +1,83 @@
+{
+    public static final SparkDate INSTANCE = new SparkDate();
+
+    private SparkDate()
```
can we annotate methods with override here?

frankgh commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/SparkInet.java:
```diff
@@ -0,0 +1,58 @@
+            {
+                return InetAddress.getByAddress((byte[]) value);
+            }
+            catch (final UnknownHostException e)
```
NIT remove final modifier
```suggestion
            catch (UnknownHostException e)
```

frankgh commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/complex/SparkMap.java:
```diff
@@ -0,0 +1,140 @@
+
+    private ArrayBasedMapData mapToSparkSqlType(final java.util.Map<Object, Object> map, boolean isFrozen)
+    {
+        final Object[] keys = new Object[map.size()];
+        final Object[] values = new Object[map.size()];
+        int position = 0;
+        for (final java.util.Map.Entry<Object, Object> entry : map.entrySet())
```
NIT Remove final qualifiers
```suggestion
        Object[] keys = new Object[map.size()];
        Object[] values = new Object[map.size()];
        int position = 0;
        for (java.util.Map.Entry<Object, Object> entry : map.entrySet())
```

frankgh commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/complex/SparkMap.java:
```diff
@@ -0,0 +1,140 @@
+    @Override
+    public Object sparkSqlRowValue(GenericInternalRow row, int position)
+    {
+        final MapData map = row.getMap(position);
+        final ArrayData keys = map.keyArray();
+        final ArrayData values = map.valueArray();
+        final java.util.Map<Object, Object> result = new HashMap<>(keys.numElements());
+        for (int element = 0; element < keys.numElements(); element++)
+        {
+            final Object key = keyType().toTestRowType(keys.get(element, keyType().dataType()));
+            final Object value = valueType().toTestRowType(values.get(element, valueType().dataType()));
```
Ditto
```suggestion
        MapData map = row.getMap(position);
        ArrayData keys = map.keyArray();
        ArrayData values = map.valueArray();
        java.util.Map<Object, Object> result = new HashMap<>(keys.numElements());
        for (int element = 0; element < keys.numElements(); element++)
        {
            Object key = keyType().toTestRowType(keys.get(element, keyType().dataType()));
            Object value = valueType().toTestRowType(values.get(element, valueType().dataType()));
```

frankgh commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/complex/SparkTuple.java:
```diff
@@ -0,0 +1,141 @@
+    @Override
+    public Object sparkSqlRowValue(GenericInternalRow row, int position)
+    {
+        final InternalRow tupleStruct = row.getStruct(position, size());
```
NIT unnecessary final modifier
```suggestion
        InternalRow tupleStruct = row.getStruct(position, size());
```

frankgh commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/complex/SparkTuple.java:
```diff
@@ -0,0 +1,141 @@
+    @Override
+    public Object sparkSqlRowValue(Row row, int position)
+    {
+        final Row tupleStruct = row.getStruct(position);
```
NIT unnecessary final modifier
```suggestion
        Row tupleStruct = row.getStruct(position);
```

frankgh commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/complex/SparkTuple.java:
```diff
@@ -0,0 +1,141 @@
+    @Override
+    public Object toTestRowType(Object value)
+    {
+        final GenericRowWithSchema tupleRow = (GenericRowWithSchema) value;
+        final Object[] tupleResult = new Object[tupleRow.size()];
```
NIT unnecessary final modifier
```suggestion
        GenericRowWithSchema tupleRow = (GenericRowWithSchema) value;
        Object[] tupleResult = new Object[tupleRow.size()];
```

frankgh commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/complex/SparkUdt.java:
```diff
@@ -0,0 +1,179 @@
+    @Override
+    public Object sparkSqlRowValue(GenericInternalRow row, int position)
+    {
+        final InternalRow struct = row.getStruct(position, size());
```
NIT unnecessary final modifier
```suggestion
        InternalRow struct = row.getStruct(position, size());
```

frankgh commented on cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/complex/SparkUdt.java:
```diff
@@ -0,0 +1,179 @@
+    @Override
+    public Object sparkSqlRowValue(Row row, int position)
+    {
+        final Row struct = row.getStruct(position);
```
NIT unnecessary final modifier
```suggestion
        Row struct = row.getStruct(position);
```

frankgh commented on cassandra-bridge/build.gradle:
```diff
@@ -40,6 +40,7 @@ publishing {
 
 dependencies {
     api(project(':cassandra-analytics-common'))
+    api(project(':cassandra-analytics-spark-converter'))
```
does this belong here? it feels the bridge should be spark agnostic. We should be able to take the bridge and use it elsewhere without having to bring the spark bits

jberragan replied:
I think bridge is the connecting point between Cassandra and Spark. It already depends on the `spark-core`, `spark_sql`  jars.

frankgh commented on cassandra-four-zero-bridge/build.gradle:
```diff
@@ -32,6 +32,7 @@ dependencies {
     compileOnly project(":cassandra-analytics-common")
     compileOnly project(":cassandra-four-zero-types")
     compileOnly(project(':cassandra-bridge'))
+    compileOnly project(":cassandra-analytics-spark-converter")
```
same question here, it feels like spark dependency does not belong here

frankgh commented on cassandra-four-zero-bridge/src/main/java/org/apache/cassandra/bridge/CassandraBridgeImplementation.java:
```diff
@@ -176,16 +176,18 @@ public CassandraBridgeImplementation()
 
+    public SparkSqlTypeConverter typeConverter()
+    {
+        return SparkSqlTypeConverterImplementation.INSTANCE;
```
this should be a param to the bridge implementation . When Spark is needed we should have this spark instance here. but if we don't need spark, we should be able to use a different impl

jberragan replied:
We could remove the Spark dependency from `cassandra-bridge` but it would be a lot more work and probably better to do outside this PR.

frankgh commented on cassandra-four-zero-bridge/src/main/java/org/apache/cassandra/spark/data/converter/SparkSqlTypeConverterImplementation.java:
```diff
@@ -0,0 +1,191 @@
+/**
+ * SparkSqlTypeConverter implementation that maps Cassandra 4.0 types to SparkSQL data types.
+ */
+public class SparkSqlTypeConverterImplementation implements SparkSqlTypeConverter
```
this feels does not belong in this project. bridges should be agnostic to spark

frankgh reviewed (commented):
Looks good in general, but I have a few comments

jberragan committed:
Review comments

jberragan committed:
Merge branch 'trunk' into CASSANDRA-19815

jberragan committed:
Remove final variables

frankgh reviewed (approved):
+1 discussed offline about removing the spark dependency and decided that it can be done as part of a follow up

jberragan committed:
Final clean up
